### ISSUE TITLE
Issue Overwriting Environment Variables

### ISSUE THREAD

**Issue by leroyjware** [CoorpAcademy/serverless-plugins#20]
What is the purpose of overwriting process.env on line 138? I've been wrestling with a pretty nasty bug for the last few days and traced it down to this single line of code. When I comment it, the function environment is preserved as defined in serverless.yml, but when I leave it uncommented, I lose all environment variables defined under provider.environment.

Some feedback would be appreciated. This is a project for a very large client and I'd love to use the plugin. It is working well otherwise with ElasticMQ running locally.

https://github.com/godu/serverless/blob/589ee1901d68f39f71a6e61682aa28018554f0be/packages/serverless-offline-sqs/src/index.js#L138

**Comment by pauldprice** [comment]
The code is messing with process.env before calling the handler and then it changes it back after the handler is invoked. However, if you have any async code that is called by the handler then that code may run after process.env is set back. I'm running into this same issue and am wondering why process.env should be reset?

**Comment by fusionbeam** [comment]
same here ...

**Comment by abbyhuisman-rw** [comment]
I'm encountering an issue related to this as well. Is there a plan to change this?

**Comment by godu** [comment]
Hi,

First of all, IMO, it's more a `serverless-offline`'s bug than `serverless-offline-{kinesis/sqs/dynamodb-stream}`'s one.

Unfortunately, I think we need this kind of hack to avoid possible side-effects produce by some `process.env`s mutations.

I'll explain.

AWS Lambda allow us to defined different environment variables by lambda. However, `serverless-offline`, by default, execute all lambdas in the same thread, and share, in the facts, the same `process.env` value.

By example, imagine two lambdas `lambda-red` and `lambda-blue` :

```yaml
functions:
  lambda-red:
    handler: handler.color
    environment:
      COLOR: red
  lambda-blue:
    handler: handler.color
    environment:
      COLOR: blue
```

with the same handler :

```js
module.exports.color = (event, context, color) => {
  console.log(process.env.COLOR);
  setTimeout(() => {
    console.log(process.env.COLOR);
    cb();
  }, 1000);
};
```

If we trigger both in  the same time, we got this output:

```shell
lambda-red: red
lambda-blue: blue
# 1s later
lambda-red: blue
lambda-blue: blue
```

A solution is to use `--useSeparateProcesses` options to isolate lambda's execution's context.

I think a good practice is to extract environment variables in the root scope, dut to the `process.env`'s immutability property that should have.

```js
const {COLOR} = process.env;
module.exports.color = (event, context, color) => {
  console.log(COLOR);
  setTimeout(() => {
    console.log(COLOR);
    cb();
  }, 1000);
};
```

I'm open for suggestion. Maybe there is another solution.

**Comment by esetnik** [comment]
@godu I think this should be reopened. `serverless-offline` is now recommending the use of `--useWorkerThreads` for node v11.7+ https://github.com/dherault/serverless-offline/issues/508#issuecomment-530877650 because there's no way to debug when using `--useSeparateProcesses`, however `--useWorkerThreads` doesn't work with `serverless-offline-sqs` in that the process.env is still wrong for the function despite being launched in a separate worker thread.

**Comment by thetumper** [comment]
Any chance to get a workaround or fix for those who cannot move to node v11.7+?  I've been looking at other options, but have not found any simple alternative.
